File: 프로그래머스/1/12921. 소수 찾기/소수 찾기.java
class Solution {
    public int solution(int n) {
        if (n < 2) {
            return 0;
        }

        // n까지의 소수를 구하기 위해 길이 n+1의 배열을 생성하고 모두 true로 초기화
        boolean[] isPrime = new boolean[n + 1];
        for (int i = 2; i <= n; i++) {
            isPrime[i] = true;
        }

        // 에라토스테네스의 체 알고리즘
        for (int p = 2; p * p <= n; p++) {
            if (isPrime[p]) {
                // p의 배수들을 false로 설정
                for (int multiple = p * p; multiple <= n; multiple += p) {
                    isPrime[multiple] = false;
                }
            }
        }

        // true인 값들의 개수를 세어 반환
        int count = 0;
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                count++;
            }
        }

        return count;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    const isPrime = num => {
        if (num <= 1) return false;
        if (num <= 3) return true;
        if (num % 2 === 0 || num % 3 === 0) return false;
        for (let i = 5; i * i <= num; i += 6) {
            if (num % i === 0 || num % (i + 2) === 0) return false;
        }
        return true;
    };

    const getPermutations = (arr, length) => {
        const results = [];
        if (length === 1) return arr.map(el => [el]);

        arr.forEach((fixed, index, origin) => {
            const rest = origin.slice(0, index).concat(origin.slice(index + 1));
            const permutations = getPermutations(rest, length - 1);
            permutations.forEach(permutation => {
                results.push([fixed].concat(permutation));
            });
        });

        return results;
    };

    const nums = new Set();
    const digits = numbers.split('');
    
    for (let length = 1; length <= digits.length; length++) {
        const permutations = getPermutations(digits, length);
        permutations.forEach(permutation => {
            const num = parseInt(permutation.join(''), 10);
            if (num > 1) nums.add(num);
        });
    }

    let count = 0;
    nums.forEach(num => {
        if (isPrime(num)) count++;
    });

    return count;
}